In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline

In [3]:
# import excel file November2008.xlsx
tn_2008=pd.read_excel('../data/2008_election/November2008.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','BNAME1','TALLY1','BNAME2','TALLY2',
                               'BNAME3','TALLY3','BNAME4','TALLY4','BNAME5','TALLY5','BNAME6','TALLY6','BNAME7','TALLY7',
                               'BNAME8','TALLY8','BNAME9','TALLY9','BNAME10','TALLY10'])
tn_2008

COUNTY            PRECINCT  \
0      ANDERSON       Andersonville   
1      ANDERSON       Andersonville   
2      ANDERSON       Andersonville   
3      ANDERSON       Andersonville   
4      ANDERSON          Briceville   
...         ...                 ...   
10609    WILSON  23 West Hills Bapt   
10610    WILSON   07 Watertown High   
10611    WILSON   07 Watertown High   
10612    WILSON   07 Watertown High   
10613    WILSON   07 Watertown High   

                                        OFFICENAME         ELECTDATE  \
0                          United States President  November 4, 2008   
1                             United States Senate  November 4, 2008   
2           U.S. House of Representatives District  November 4, 2008   
3      Tennessee House of Representatives District  November 4, 2008   
4                          United States President  November 4, 2008   
...                                            ...               ...   
10609  Tennessee House of Representatives District  November 4, 2008   
10610                      United States President  November 4, 2008   
10611                         United States Senate  November 4, 2008   
10612       U.S. House of Representatives District  November 4, 2008   
10613  Tennessee House of Representatives District  November 4, 2008   

                        BNAME1  TALLY1                 BNAME2  TALLY2  \
0           Barack Obama - (D)     290      John McCain - (R)    1004   
1         Robert D. Tuke - (D)     221  Lamar Alexander - (R)     989   
2        Doug Vandagriff - (D)     229        Zach Wamp - (R)     974   
3      Leslie Winningham - (D)     709                    NaN       0   
4           Barack Obama - (D)     120      John McCain - (R)     198   
...                        ...     ...                    ...     ...   
10609      Stratton Bone - (D)     834    Albert McCall - (R)     817   
10610       Barack Obama - (D)     152      John McCain - (R)     306   
10611     Robert D. Tuke - (D)     110  Lamar Alexander - (R)     291   
10612        Bart Gordon - (D)     282      Chris Baker - (I)     105   
10613      Stratton Bone - (D)     244    Albert McCall - (R)     203   

                     BNAME3  TALLY3  ...                  BNAME6  TALLY6  \
0       Chuck Baldwin - (I)       3  ...  Cynthia McKinney - (I)       0   
1      Edward L. Buck - (I)      13  ...        Ed Lawhorn - (I)       6   
2           Ed Choate - (I)      15  ...                     NaN       0   
3                       NaN       0  ...                     NaN       0   
4       Chuck Baldwin - (I)       2  ...  Cynthia McKinney - (I)       0   
...                     ...     ...  ...                     ...     ...   
10609                   NaN       0  ...                     NaN       0   
10610   Chuck Baldwin - (I)       2  ...  Cynthia McKinney - (I)       0   
10611  Edward L. Buck - (I)      15  ...        Ed Lawhorn - (I)       3   
10612                   NaN       0  ...                     NaN       0   
10613                   NaN       0  ...                     NaN       0   

                          BNAME7  TALLY7             BNAME8  TALLY8  \
0              Brian Moore - (I)       1  Ralph Nader - (I)       7   
1      Daniel Towers Lewis - (I)       3   Chris Lugo - (I)       5   
2                            NaN       0                NaN       0   
3                            NaN       0                NaN       0   
4              Brian Moore - (I)       0  Ralph Nader - (I)       1   
...                          ...     ...                ...     ...   
10609                        NaN       0                NaN       0   
10610          Brian Moore - (I)       0  Ralph Nader - (I)       2   
10611  Daniel Towers Lewis - (I)       2   Chris Lugo - (I)       0   
10612                        NaN       0                NaN       0   
10613                        NaN       0                NaN       0   

                       BNAME9  T

In [4]:
tn_2008.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10614 entries, 0 to 10613
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      10614 non-null  object 
 1   PRECINCT    10614 non-null  object 
 2   OFFICENAME  10614 non-null  object 
 3   ELECTDATE   10614 non-null  object 
 4   BNAME1      10614 non-null  object 
 5   TALLY1      10614 non-null  int64  
 6   BNAME2      9005 non-null   object 
 7   TALLY2      10614 non-null  int64  
 8   BNAME3      6367 non-null   object 
 9   TALLY3      10614 non-null  int64  
 10  BNAME4      6096 non-null   object 
 11  TALLY4      10614 non-null  int64  
 12  BNAME5      5313 non-null   object 
 13  TALLY5      10614 non-null  int64  
 14  BNAME6      4794 non-null   object 
 15  TALLY6      10614 non-null  int64  
 16  BNAME7      4794 non-null   object 
 17  TALLY7      10614 non-null  int64  
 18  BNAME8      4592 non-null   object 
 19  TALLY8      10614 non-nul

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2008_namemelt=tn_2008.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['BNAME1','BNAME2','BNAME3','BNAME4','BNAME5','BNAME6','BNAME7','BNAME8','BNAME9','BNAME10'],value_name="candidate")
# drop column "variable" created by .melt
tn_2008_namemelt=tn_2008_namemelt.drop(['variable'],axis=1)                          
tn_2008_namemelt.head()

COUNTY       PRECINCT                                   OFFICENAME  \
0  ANDERSON  Andersonville                      United States President   
1  ANDERSON  Andersonville                         United States Senate   
2  ANDERSON  Andersonville       U.S. House of Representatives District   
3  ANDERSON  Andersonville  Tennessee House of Representatives District   
4  ANDERSON     Briceville                      United States President   

          ELECTDATE                candidate  
0  November 4, 2008       Barack Obama - (D)  
1  November 4, 2008     Robert D. Tuke - (D)  
2  November 4, 2008    Doug Vandagriff - (D)  
3  November 4, 2008  Leslie Winningham - (D)  
4  November 4, 2008       Barack Obama - (D)

In [6]:
# merge entries in all TALLY columns into new column "vote_count"
tn_2008_pvtallymelt=tn_2008.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['TALLY1','TALLY2','TALLY3','TALLY4','TALLY5','TALLY6','TALLY7','TALLY8','TALLY9','TALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2008_pvtallymelt=tn_2008_pvtallymelt.drop(['variable'],axis=1)
tn_2008_pvtallymelt.head()

COUNTY       PRECINCT                                   OFFICENAME  \
0  ANDERSON  Andersonville                      United States President   
1  ANDERSON  Andersonville                         United States Senate   
2  ANDERSON  Andersonville       U.S. House of Representatives District   
3  ANDERSON  Andersonville  Tennessee House of Representatives District   
4  ANDERSON     Briceville                      United States President   

          ELECTDATE  vote_count  
0  November 4, 2008         290  
1  November 4, 2008         221  
2  November 4, 2008         229  
3  November 4, 2008         709  
4  November 4, 2008         120

In [7]:
# add "vote_count" to tn_2008_namemelt
tally=tn_2008_pvtallymelt['vote_count']
tn_2008_namemelt=tn_2008_namemelt.join(tally)
tn_2008_namemelt.head(3)

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE              candidate  vote_count  
0  November 4, 2008     Barack Obama - (D)         290  
1  November 4, 2008   Robert D. Tuke - (D)         221  
2  November 4, 2008  Doug Vandagriff - (D)         229

In [8]:
tn_2008_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106140 entries, 0 to 106139
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   COUNTY      106140 non-null  object
 1   PRECINCT    106140 non-null  object
 2   OFFICENAME  106140 non-null  object
 3   ELECTDATE   106140 non-null  object
 4   candidate   53871 non-null   object
 5   vote_count  106140 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 4.9+ MB


In [9]:
# lower case column names
tn_2008_namemelt.columns=tn_2008_namemelt.columns.str.lower()
tn_2008_namemelt.head(2)

county       precinct               officename         electdate  \
0  ANDERSON  Andersonville  United States President  November 4, 2008   
1  ANDERSON  Andersonville     United States Senate  November 4, 2008   

              candidate  vote_count  
0    Barack Obama - (D)         290  
1  Robert D. Tuke - (D)         221

In [10]:
tn_2008_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106140 entries, 0 to 106139
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      106140 non-null  object
 1   precinct    106140 non-null  object
 2   officename  106140 non-null  object
 3   electdate   106140 non-null  object
 4   candidate   53871 non-null   object
 5   vote_count  106140 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 4.9+ MB


#### 53871 non-null names  out of 106140 enties. Confirm all are 0 votes and drop rows with no candidate listed

In [11]:
# Create dataframe "blank_candidate_2008" of null valuea in "candidate" of tn_2008_namemelt
blank_candidate_2008=tn_2008_namemelt[tn_2008_namemelt['candidate'].isna()]
blank_candidate_2008

county            precinct  \
10617   ANDERSON       Andersonville   
10621   ANDERSON          Briceville   
10677   ANDERSON           Lake City   
10681   ANDERSON    Lake City Middle   
10697   ANDERSON             Norwood   
...          ...                 ...   
106135    WILSON  23 West Hills Bapt   
106136    WILSON   07 Watertown High   
106137    WILSON   07 Watertown High   
106138    WILSON   07 Watertown High   
106139    WILSON   07 Watertown High   

                                         officename         electdate  \
10617   Tennessee House of Representatives District  November 4, 2008   
10621   Tennessee House of Representatives District  November 4, 2008   
10677   Tennessee House of Representatives District  November 4, 2008   
10681   Tennessee House of Representatives District  November 4, 2008   
10697   Tennessee House of Representatives District  November 4, 2008   
...                                             ...               ...   
106135  Tennessee House of Representatives District  November 4, 2008   
106136                      United States President  November 4, 2008   
106137                         United States Senate  November 4, 2008   
106138       U.S. House of Representatives District  November 4, 2008   
106139  Tennessee House of Representatives District  November 4, 2008   

       candidate  vote_count  
10617        NaN           0  
10621        NaN           0  
10677        NaN           0  
10681        NaN           0  
10697        NaN           0  
...          ...         ...  
106135       NaN           0  
106136       NaN           0  
106137       NaN           0  
106138       NaN           0  
106139       NaN           0  

[52269 rows x 6 columns]

In [12]:
# confirm all are 0 counts
blank_candidate_2008['vote_count'].value_counts()

0    52269
Name: vote_count, dtype: int64

In [13]:
# drop all null candidates
tn_2008_namemelt.dropna().reset_index(drop=True)

county            precinct  \
0      ANDERSON       Andersonville   
1      ANDERSON       Andersonville   
2      ANDERSON       Andersonville   
3      ANDERSON       Andersonville   
4      ANDERSON          Briceville   
...         ...                 ...   
53866    WILSON  21 Water Authority   
53867    WILSON  22 W A Wright Elem   
53868    WILSON   09 Watertown Comm   
53869    WILSON  23 West Hills Bapt   
53870    WILSON   07 Watertown High   

                                        officename         electdate  \
0                          United States President  November 4, 2008   
1                             United States Senate  November 4, 2008   
2           U.S. House of Representatives District  November 4, 2008   
3      Tennessee House of Representatives District  November 4, 2008   
4                          United States President  November 4, 2008   
...                                            ...               ...   
53866                         United States Senate  November 4, 2008   
53867                         United States Senate  November 4, 2008   
53868                         United States Senate  November 4, 2008   
53869                         United States Senate  November 4, 2008   
53870                         United States Senate  November 4, 2008   

                     candidate  vote_count  
0           Barack Obama - (D)         290  
1         Robert D. Tuke - (D)         221  
2        Doug Vandagriff - (D)         229  
3      Leslie Winningham - (D)         709  
4           Barack Obama - (D)         120  
...                        ...         ...  
53866   Write-In - Casey Gouge           0  
53867   Write-In - Casey Gouge           0  
53868   Write-In - Casey Gouge           0  
53869   Write-In - Casey Gouge           0  
53870   Write-In - Casey Gouge           0  

[53871 rows x 6 columns]

In [14]:
# confirn all null candidates dropped
tn_2008_namemelt[tn_2008_namemelt['candidate'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52269 entries, 10617 to 106139
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      52269 non-null  object
 1   precinct    52269 non-null  object
 2   officename  52269 non-null  object
 3   electdate   52269 non-null  object
 4   candidate   0 non-null      object
 5   vote_count  52269 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 2.8+ MB


In [15]:
tn_2008_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106140 entries, 0 to 106139
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      106140 non-null  object
 1   precinct    106140 non-null  object
 2   officename  106140 non-null  object
 3   electdate   106140 non-null  object
 4   candidate   53871 non-null   object
 5   vote_count  106140 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 4.9+ MB


### Look at cadidates with 0 votes, drop

In [16]:
tn_2008_namemelt_0votes=tn_2008_namemelt[tn_2008_namemelt['vote_count'] == 0]
tn_2008_namemelt_0votes

county             precinct  \
2005    DAVIDSON  20-2 Union Hall 737   
3459    HAMILTON        112 Airport 2   
3460    HAMILTON        112 Airport 2   
3461    HAMILTON        112 Airport 2   
3462    HAMILTON        112 Airport 2   
...          ...                  ...   
106135    WILSON   23 West Hills Bapt   
106136    WILSON    07 Watertown High   
106137    WILSON    07 Watertown High   
106138    WILSON    07 Watertown High   
106139    WILSON    07 Watertown High   

                                         officename         electdate  \
2005    Tennessee House of Representatives District  November 4, 2008   
3459                        United States President  November 4, 2008   
3460                           United States Senate  November 4, 2008   
3461         U.S. House of Representatives District  November 4, 2008   
3462    Tennessee House of Representatives District  November 4, 2008   
...                                             ...               ...   
106135  Tennessee House of Representatives District  November 4, 2008   
106136                      United States President  November 4, 2008   
106137                         United States Senate  November 4, 2008   
106138       U.S. House of Representatives District  November 4, 2008   
106139  Tennessee House of Representatives District  November 4, 2008   

                     candidate  vote_count  
2005       Gary W. Moore - (D)           0  
3459        Barack Obama - (D)           0  
3460      Robert D. Tuke - (D)           0  
3461     Doug Vandagriff - (D)           0  
3462    Gerald McCormick - (R)           0  
...                        ...         ...  
106135                     NaN           0  
106136                     NaN           0  
106137                     NaN           0  
106138                     NaN           0  
106139                     NaN           0  

[63386 rows x 6 columns]

In [17]:
# drop rows with 0 in "vote_count" column
tn_2008_namemelt=tn_2008_namemelt[tn_2008_namemelt['vote_count'] != 0].reset_index(drop=True)
tn_2008_namemelt

county            precinct  \
0      ANDERSON       Andersonville   
1      ANDERSON       Andersonville   
2      ANDERSON       Andersonville   
3      ANDERSON       Andersonville   
4      ANDERSON          Briceville   
...         ...                 ...   
42749    WILSON   09 Watertown Comm   
42750    WILSON   09 Watertown Comm   
42751    WILSON  23 West Hills Bapt   
42752    WILSON  23 West Hills Bapt   
42753    WILSON   07 Watertown High   

                                        officename         electdate  \
0                          United States President  November 4, 2008   
1                             United States Senate  November 4, 2008   
2           U.S. House of Representatives District  November 4, 2008   
3      Tennessee House of Representatives District  November 4, 2008   
4                          United States President  November 4, 2008   
...                                            ...               ...   
42749                      United States President  November 4, 2008   
42750                         United States Senate  November 4, 2008   
42751                      United States President  November 4, 2008   
42752                         United States Senate  November 4, 2008   
42753                      United States President  November 4, 2008   

                     candidate  vote_count  
0           Barack Obama - (D)         290  
1         Robert D. Tuke - (D)         221  
2        Doug Vandagriff - (D)         229  
3      Leslie Winningham - (D)         709  
4           Barack Obama - (D)         120  
...                        ...         ...  
42749        Ralph Nader - (I)           4  
42750         Chris Lugo - (I)           6  
42751        Ralph Nader - (I)          12  
42752         Chris Lugo - (I)           4  
42753        Ralph Nader - (I)           2  

[42754 rows x 6 columns]

In [18]:
tn_2008_namemelt['vote_count'].value_counts()

1       5128
2       3500
3       2684
4       2155
5       1658
        ... 
2537       1
2103       1
2473       1
2231       1
3414       1
Name: vote_count, Length: 2308, dtype: int64

In [19]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2008_int=tn_2008_namemelt.groupby(['electdate','officename','county','candidate'], as_index=False)['vote_count'].sum()
tn_2008_int=tn_2008_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2008_int.head(30)

electdate                                   officename      county  \
0   November 4, 2008                         United States Senate      WILSON   
1   November 4, 2008                      United States President      WILSON   
2   November 4, 2008  Tennessee House of Representatives District      WILSON   
3   November 4, 2008       U.S. House of Representatives District      WILSON   
4   November 4, 2008       U.S. House of Representatives District      WILSON   
5   November 4, 2008                      United States President      WILSON   
6   November 4, 2008                         United States Senate      WILSON   
7   November 4, 2008  Tennessee House of Representatives District      WILSON   
8   November 4, 2008  Tennessee House of Representatives District      WILSON   
9   November 4, 2008  Tennessee House of Representatives District      WILSON   
10  November 4, 2008       U.S. House of Representatives District      WILSON   
11  November 4, 2008       U.S. House of Representatives District      WILSON   
12  November 4, 2008                         United States Senate      WILSON   
13  November 4, 2008       U.S. House of Representatives District      WILSON   
14  November 4, 2008       U.S. House of Representatives District      WILSON   
15  November 4, 2008                         United States Senate      WILSON   
16  November 4, 2008                      United States President      WILSON   
17  November 4, 2008                         United States Senate      WILSON   
18  November 4, 2008                         United States Senate      WILSON   
19  November 4, 2008                         United States Senate      WILSON   
20  November 4, 2008                      United States President      WILSON   
21  November 4, 2008                      United States President      WILSON   
22  November 4, 2008                         United States Senate      WILSON   
23  November 4, 2008                      United States President      WILSON   
24  November 4, 2008                      United States President      WILSON   
25  November 4, 2008                      United States President      WILSON   
26  November 4, 2008                         United States Senate  WILLIAMSON   
27  November 4, 2008                      United States President  WILLIAMSON   
28  November 4, 2008       U.S. House of Representatives District  WILLIAMSON   
29  November 4, 2008  Tennessee House of Representatives District  WILLIAMSON   

                      candidate  vote_count  
0         Lamar Alexander - (R)       34781  
1             John McCain - (R)       34595  
2           Susan M. Lynn - (R)       20959  
3          Gerard Donovan - (R)       19866  
4              Jim Cooper - (D)       17471  
5            Barack Obama - (D)       15886  
6          Robert D. Tuke - (D)       12240  
7           Stratton Bone - (D)        9252  
8           Ken Wilkinson - (D)        8386  
9           Albert McCall - (R)        8284  
10            Bart Gordon - (D)        6415  
11            Chris Baker - (I)        2441  
12         Edward L. Buck - (I)        1127  
13            Jon Jackson - (I)         751  
14      John P. Miglietta - (I)         294  
15  Christopher G. Fenner - (I)         230  
16            Ralph Nader - (I)         226  
17    Daniel Towers Lewis - (I)         218  
18             Ed Lawhorn - (I)         207  
19             Chris Lugo - (I)         195  
20          Chuck Baldwin - (I)         186  
21               Bob Barr - (I)         174  
22         David Gatchell - (I)         145  
23       Cynthia McKinney - (I)          45  
24            Brian Moore - (I)          32  
25            Charles Jay - (I)          15  
26        Lamar Alexander - (R)       68258  
27            John McCain - (R)       64858  
28       Marsha Blackburn - (R)       56674  
29            Glen Casada - (R)       37948

In [20]:
# Capitalize couty names
tn_2008_int['county']=tn_2008_int['county'].str.title()
tn_2008_int.head(5)

electdate                                   officename  county  \
0  November 4, 2008                         United States Senate  Wilson   
1  November 4, 2008                      United States President  Wilson   
2  November 4, 2008  Tennessee House of Representatives District  Wilson   
3  November 4, 2008       U.S. House of Representatives District  Wilson   
4  November 4, 2008       U.S. House of Representatives District  Wilson   

               candidate  vote_count  
0  Lamar Alexander - (R)       34781  
1      John McCain - (R)       34595  
2    Susan M. Lynn - (R)       20959  
3   Gerard Donovan - (R)       19866  
4       Jim Cooper - (D)       17471

In [21]:
tn_2008_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2194 entries, 0 to 2193
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   electdate   2194 non-null   object
 1   officename  2194 non-null   object
 2   county      2194 non-null   object
 3   candidate   2194 non-null   object
 4   vote_count  2194 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 85.8+ KB


In [22]:
tn_2008_int['officename'].unique()

array(['United States Senate', 'United States President',
       'Tennessee House of Representatives District',
       'U.S. House of Representatives District',
       'Tennessee Senate District'], dtype=object)

### Clean up names and greate separate party column

In [23]:
# new dataframe new_names of non-write-in candidates
new_names=tn_2008_int[tn_2008_int['candidate'].str.contains("Write")==False]
new_names

electdate                                   officename    county  \
0     November 4, 2008                         United States Senate    Wilson   
1     November 4, 2008                      United States President    Wilson   
2     November 4, 2008  Tennessee House of Representatives District    Wilson   
3     November 4, 2008       U.S. House of Representatives District    Wilson   
4     November 4, 2008       U.S. House of Representatives District    Wilson   
...                ...                                          ...       ...   
2189  November 4, 2008                         United States Senate  Anderson   
2190  November 4, 2008                         United States Senate  Anderson   
2191  November 4, 2008                      United States President  Anderson   
2192  November 4, 2008                      United States President  Anderson   
2193  November 4, 2008                      United States President  Anderson   

                   candidate  vote_count  
0      Lamar Alexander - (R)       34781  
1          John McCain - (R)       34595  
2        Susan M. Lynn - (R)       20959  
3       Gerard Donovan - (R)       19866  
4           Jim Cooper - (D)       17471  
...                      ...         ...  
2189        Chris Lugo - (I)          90  
2190    David Gatchell - (I)          84  
2191  Cynthia McKinney - (I)          24  
2192       Brian Moore - (I)          11  
2193       Charles Jay - (I)           9  

[2179 rows x 5 columns]

In [24]:
# split name fromm party desination
new_names=new_names['candidate'].str.split(pat='-',expand=True)
new_names

0     1     2
0      Lamar Alexander    (R)  None
1          John McCain    (R)  None
2        Susan M. Lynn    (R)  None
3       Gerard Donovan    (R)  None
4           Jim Cooper    (D)  None
...                 ...   ...   ...
2189        Chris Lugo    (I)  None
2190    David Gatchell    (I)  None
2191  Cynthia McKinney    (I)  None
2192       Brian Moore    (I)  None
2193       Charles Jay    (I)  None

[2179 rows x 3 columns]

In [25]:
new_names.rename(columns={0:'name',1:'party'},inplace=True)
new_names=new_names.drop([2],axis=1)
new_names

name party
0      Lamar Alexander    (R)
1          John McCain    (R)
2        Susan M. Lynn    (R)
3       Gerard Donovan    (R)
4           Jim Cooper    (D)
...                 ...   ...
2189        Chris Lugo    (I)
2190    David Gatchell    (I)
2191  Cynthia McKinney    (I)
2192       Brian Moore    (I)
2193       Charles Jay    (I)

[2179 rows x 2 columns]

In [26]:
# new dataframe tn_2012_int_wi of only write-in candidates
tn_2008_int_wi=tn_2008_int[tn_2008_int['candidate'].str.contains("\(")==False]
tn_2008_int_wi

electdate                                   officename  \
91    November 4, 2008       U.S. House of Representatives District   
346   November 4, 2008  Tennessee House of Representatives District   
350   November 4, 2008       U.S. House of Representatives District   
597   November 4, 2008       U.S. House of Representatives District   
774   November 4, 2008                    Tennessee Senate District   
1105  November 4, 2008       U.S. House of Representatives District   
1236  November 4, 2008                    Tennessee Senate District   
1246  November 4, 2008       U.S. House of Representatives District   
1285  November 4, 2008  Tennessee House of Representatives District   
1293  November 4, 2008       U.S. House of Representatives District   
1599  November 4, 2008       U.S. House of Representatives District   
1947  November 4, 2008                    Tennessee Senate District   
1968  November 4, 2008       U.S. House of Representatives District   
2144  November 4, 2008  Tennessee House of Representatives District   
2146  November 4, 2008       U.S. House of Representatives District   

          county                    candidate  vote_count  
91       Weakley        Write-In - James Hart           9  
346      Stewart        Write-In - James Hart          24  
350      Stewart        Write-In - James Hart          11  
597         Rhea      Write-In - June Griffin           7  
774   Montgomery   Write-In - Rosalind Kurita       15174  
1105        Lake        Write-In - James Hart           1  
1236     Houston   Write-In - Rosalind Kurita        1079  
1246     Houston        Write-In - James Hart           4  
1285       Henry        Write-In - James Hart          74  
1293       Henry        Write-In - James Hart          19  
1599      Gibson        Write-In - James Hart           1  
1947    Cheatham   Write-In - Rosalind Kurita        6552  
1968    Cheatham  Write-In - Thomas F. Kovach           4  
2144      Benton        Write-In - James Hart          14  
2146      Benton        Write-In - James Hart           9

In [27]:
# split name fromm 'Write-in'
new_wi_name=tn_2008_int_wi['candidate'].str.split(pat='-',n=2,expand=True)
new_wi_name

0    1                  2
91    Write  In          James Hart
346   Write  In          James Hart
350   Write  In          James Hart
597   Write  In        June Griffin
774   Write  In     Rosalind Kurita
1105  Write  In          James Hart
1236  Write  In     Rosalind Kurita
1246  Write  In          James Hart
1285  Write  In          James Hart
1293  Write  In          James Hart
1599  Write  In          James Hart
1947  Write  In     Rosalind Kurita
1968  Write  In    Thomas F. Kovach
2144  Write  In          James Hart
2146  Write  In          James Hart

In [28]:
# create new 'party' columnn with "Write-In"
new_wi_name['party']=new_wi_name[[0,1]].agg('-'.join,axis=1)
new_wi_name

0    1                  2      party
91    Write  In          James Hart  Write-In 
346   Write  In          James Hart  Write-In 
350   Write  In          James Hart  Write-In 
597   Write  In        June Griffin  Write-In 
774   Write  In     Rosalind Kurita  Write-In 
1105  Write  In          James Hart  Write-In 
1236  Write  In     Rosalind Kurita  Write-In 
1246  Write  In          James Hart  Write-In 
1285  Write  In          James Hart  Write-In 
1293  Write  In          James Hart  Write-In 
1599  Write  In          James Hart  Write-In 
1947  Write  In     Rosalind Kurita  Write-In 
1968  Write  In    Thomas F. Kovach  Write-In 
2144  Write  In          James Hart  Write-In 
2146  Write  In          James Hart  Write-In

In [29]:
#drop columns 0 and 1, rename column 2 to"candidate"
new_wi_name=new_wi_name.drop([0,1],axis=1)
new_wi_name

2      party
91           James Hart  Write-In 
346          James Hart  Write-In 
350          James Hart  Write-In 
597        June Griffin  Write-In 
774     Rosalind Kurita  Write-In 
1105         James Hart  Write-In 
1236    Rosalind Kurita  Write-In 
1246         James Hart  Write-In 
1285         James Hart  Write-In 
1293         James Hart  Write-In 
1599         James Hart  Write-In 
1947    Rosalind Kurita  Write-In 
1968   Thomas F. Kovach  Write-In 
2144         James Hart  Write-In 
2146         James Hart  Write-In

In [30]:
new_wi_name.rename(columns={2:'name'},inplace=True)
new_wi_name

name      party
91           James Hart  Write-In 
346          James Hart  Write-In 
350          James Hart  Write-In 
597        June Griffin  Write-In 
774     Rosalind Kurita  Write-In 
1105         James Hart  Write-In 
1236    Rosalind Kurita  Write-In 
1246         James Hart  Write-In 
1285         James Hart  Write-In 
1293         James Hart  Write-In 
1599         James Hart  Write-In 
1947    Rosalind Kurita  Write-In 
1968   Thomas F. Kovach  Write-In 
2144         James Hart  Write-In 
2146         James Hart  Write-In

In [31]:
# join new_names and new_wi_names
frames=[new_names,new_wi_name]
cleaned_names=pd.concat(frames)
cleaned_names

name      party
0      Lamar Alexander         (R)
1          John McCain         (R)
2        Susan M. Lynn         (R)
3       Gerard Donovan         (R)
4           Jim Cooper         (D)
...                 ...        ...
1599         James Hart  Write-In 
1947    Rosalind Kurita  Write-In 
1968   Thomas F. Kovach  Write-In 
2144         James Hart  Write-In 
2146         James Hart  Write-In 

[2194 rows x 2 columns]

### Join cleaned names and party column to tn_2008_int

In [32]:
name=cleaned_names['name']
party=cleaned_names['party']
tn_2008_final=tn_2008_int.join(name).join(party)
tn_2008_final

electdate                                   officename    county  \
0     November 4, 2008                         United States Senate    Wilson   
1     November 4, 2008                      United States President    Wilson   
2     November 4, 2008  Tennessee House of Representatives District    Wilson   
3     November 4, 2008       U.S. House of Representatives District    Wilson   
4     November 4, 2008       U.S. House of Representatives District    Wilson   
...                ...                                          ...       ...   
2189  November 4, 2008                         United States Senate  Anderson   
2190  November 4, 2008                         United States Senate  Anderson   
2191  November 4, 2008                      United States President  Anderson   
2192  November 4, 2008                      United States President  Anderson   
2193  November 4, 2008                      United States President  Anderson   

                   candidate  vote_count               name party  
0      Lamar Alexander - (R)       34781   Lamar Alexander    (R)  
1          John McCain - (R)       34595       John McCain    (R)  
2        Susan M. Lynn - (R)       20959     Susan M. Lynn    (R)  
3       Gerard Donovan - (R)       19866    Gerard Donovan    (R)  
4           Jim Cooper - (D)       17471        Jim Cooper    (D)  
...                      ...         ...                ...   ...  
2189        Chris Lugo - (I)          90        Chris Lugo    (I)  
2190    David Gatchell - (I)          84    David Gatchell    (I)  
2191  Cynthia McKinney - (I)          24  Cynthia McKinney    (I)  
2192       Brian Moore - (I)          11       Brian Moore    (I)  
2193       Charles Jay - (I)           9       Charles Jay    (I)  

[2194 rows x 7 columns]

In [33]:
tn_2008_final['party'].value_counts()

 (I)         1290
 (D)          459
 (R)          419
Write-In       15
Hill           11
Name: party, dtype: int64

In [34]:
# change "(R)" to "Republican" and "(D)"to "Democratic" in "party" column
tn_2008_final["party"]=tn_2008_final['party'].str.replace('\(R\)',"Republican").str.replace('\(D\)',"Democratic")
tn_2008_final

electdate                                   officename    county  \
0     November 4, 2008                         United States Senate    Wilson   
1     November 4, 2008                      United States President    Wilson   
2     November 4, 2008  Tennessee House of Representatives District    Wilson   
3     November 4, 2008       U.S. House of Representatives District    Wilson   
4     November 4, 2008       U.S. House of Representatives District    Wilson   
...                ...                                          ...       ...   
2189  November 4, 2008                         United States Senate  Anderson   
2190  November 4, 2008                         United States Senate  Anderson   
2191  November 4, 2008                      United States President  Anderson   
2192  November 4, 2008                      United States President  Anderson   
2193  November 4, 2008                      United States President  Anderson   

                   candidate  vote_count               name        party  
0      Lamar Alexander - (R)       34781   Lamar Alexander    Republican  
1          John McCain - (R)       34595       John McCain    Republican  
2        Susan M. Lynn - (R)       20959     Susan M. Lynn    Republican  
3       Gerard Donovan - (R)       19866    Gerard Donovan    Republican  
4           Jim Cooper - (D)       17471        Jim Cooper    Democratic  
...                      ...         ...                ...          ...  
2189        Chris Lugo - (I)          90        Chris Lugo           (I)  
2190    David Gatchell - (I)          84    David Gatchell           (I)  
2191  Cynthia McKinney - (I)          24  Cynthia McKinney           (I)  
2192       Brian Moore - (I)          11       Brian Moore           (I)  
2193       Charles Jay - (I)           9       Charles Jay           (I)  

[2194 rows x 7 columns]

In [36]:
# create data frame tn_2012_uss for senate election
tn_2008_uss=tn_2008_final[tn_2008_final['officename'].str.match('United States Senate')]
tn_2008_uss=tn_2008_uss.reset_index(drop=True)
tn_2008_uss

electdate            officename    county  \
0    November 4, 2008  United States Senate    Wilson   
1    November 4, 2008  United States Senate    Wilson   
2    November 4, 2008  United States Senate    Wilson   
3    November 4, 2008  United States Senate    Wilson   
4    November 4, 2008  United States Senate    Wilson   
..                ...                   ...       ...   
755  November 4, 2008  United States Senate  Anderson   
756  November 4, 2008  United States Senate  Anderson   
757  November 4, 2008  United States Senate  Anderson   
758  November 4, 2008  United States Senate  Anderson   
759  November 4, 2008  United States Senate  Anderson   

                       candidate  vote_count                    name  \
0          Lamar Alexander - (R)       34781        Lamar Alexander    
1           Robert D. Tuke - (D)       12240         Robert D. Tuke    
2           Edward L. Buck - (I)        1127         Edward L. Buck    
3    Christopher G. Fenner - (I)         230  Christopher G. Fenner    
4      Daniel Towers Lewis - (I)         218    Daniel Towers Lewis    
..                           ...         ...                     ...   
755             Ed Lawhorn - (I)         151             Ed Lawhorn    
756    Daniel Towers Lewis - (I)         118    Daniel Towers Lewis    
757  Christopher G. Fenner - (I)         117  Christopher G. Fenner    
758             Chris Lugo - (I)          90             Chris Lugo    
759         David Gatchell - (I)          84         David Gatchell    

           party  
0     Republican  
1     Democratic  
2            (I)  
3            (I)  
4            (I)  
..           ...  
755          (I)  
756          (I)  
757          (I)  
758          (I)  
759          (I)  

[760 rows x 7 columns]

In [37]:
tn_2008_uss.head(20)

electdate            officename      county  \
0   November 4, 2008  United States Senate      Wilson   
1   November 4, 2008  United States Senate      Wilson   
2   November 4, 2008  United States Senate      Wilson   
3   November 4, 2008  United States Senate      Wilson   
4   November 4, 2008  United States Senate      Wilson   
5   November 4, 2008  United States Senate      Wilson   
6   November 4, 2008  United States Senate      Wilson   
7   November 4, 2008  United States Senate      Wilson   
8   November 4, 2008  United States Senate  Williamson   
9   November 4, 2008  United States Senate  Williamson   
10  November 4, 2008  United States Senate  Williamson   
11  November 4, 2008  United States Senate  Williamson   
12  November 4, 2008  United States Senate  Williamson   
13  November 4, 2008  United States Senate  Williamson   
14  November 4, 2008  United States Senate  Williamson   
15  November 4, 2008  United States Senate  Williamson   
16  November 4, 2008  United States Senate       White   
17  November 4, 2008  United States Senate       White   
18  November 4, 2008  United States Senate       White   
19  November 4, 2008  United States Senate       White   

                      candidate  vote_count                    name  \
0         Lamar Alexander - (R)       34781        Lamar Alexander    
1          Robert D. Tuke - (D)       12240         Robert D. Tuke    
2          Edward L. Buck - (I)        1127         Edward L. Buck    
3   Christopher G. Fenner - (I)         230  Christopher G. Fenner    
4     Daniel Towers Lewis - (I)         218    Daniel Towers Lewis    
5              Ed Lawhorn - (I)         207             Ed Lawhorn    
6              Chris Lugo - (I)         195             Chris Lugo    
7          David Gatchell - (I)         145         David Gatchell    
8         Lamar Alexander - (R)       68258        Lamar Alexander    
9          Robert D. Tuke - (D)       19405         Robert D. Tuke    
10         Edward L. Buck - (I)        1262         Edward L. Buck    
11             Chris Lugo - (I)         342             Chris Lugo    
12    Daniel Towers Lewis - (I)         331    Daniel Towers Lewis    
13  Christopher G. Fenner - (I)         289  Christopher G. Fenner    
14             Ed Lawhorn - (I)         277             Ed Lawhorn    
15         David Gatchell - (I)         258         David Gatchell    
16        Lamar Alexander - (R)        5969        Lamar Alexander    
17         Robert D. Tuke - (D)        2620         Robert D. Tuke    
18         Edward L. Buck - (I)         231         Edward L. Buck    
19  Christopher G. Fenner - (I)          54  Christopher G. Fenner    

          party  
0    Republican  
1    Democratic  
2           (I)  
3           (I)  
4           (I)  
5           (I)  
6           (I)  
7           (I)  
8    Republican  
9    Democratic  
10          (I)  
11          (I)  
12          (I)  
13          (I)  
14          (I)  
15          (I)  
16   Republican  
17   Democratic  
18          (I)  
19          (I)

### Add column that has total vote count by county for use in calculating county vote percentages

In [38]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2008_uss.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       34781
1      Wilson       12240
2      Wilson        1127
3      Wilson         230
4      Wilson         218
..        ...         ...
755  Anderson         151
756  Anderson         118
757  Anderson         117
758  Anderson          90
759  Anderson          84

[760 rows x 2 columns]

In [39]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      29923
Bedford       13808
Benton         6053
Bledsoe        4508
Blount        49218
              ...  
Wayne          4837
Weakley       12814
White          9016
Williamson    90422
Wilson        49143
Name: vote_count, Length: 95, dtype: int64

In [40]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2008_uss['sum_county_vote']=tn_2008_uss['county'].map(county_vote_sum)
tn_2008_uss.head(20)

electdate            officename      county  \
0   November 4, 2008  United States Senate      Wilson   
1   November 4, 2008  United States Senate      Wilson   
2   November 4, 2008  United States Senate      Wilson   
3   November 4, 2008  United States Senate      Wilson   
4   November 4, 2008  United States Senate      Wilson   
5   November 4, 2008  United States Senate      Wilson   
6   November 4, 2008  United States Senate      Wilson   
7   November 4, 2008  United States Senate      Wilson   
8   November 4, 2008  United States Senate  Williamson   
9   November 4, 2008  United States Senate  Williamson   
10  November 4, 2008  United States Senate  Williamson   
11  November 4, 2008  United States Senate  Williamson   
12  November 4, 2008  United States Senate  Williamson   
13  November 4, 2008  United States Senate  Williamson   
14  November 4, 2008  United States Senate  Williamson   
15  November 4, 2008  United States Senate  Williamson   
16  November 4, 2008  United States Senate       White   
17  November 4, 2008  United States Senate       White   
18  November 4, 2008  United States Senate       White   
19  November 4, 2008  United States Senate       White   

                      candidate  vote_count                    name  \
0         Lamar Alexander - (R)       34781        Lamar Alexander    
1          Robert D. Tuke - (D)       12240         Robert D. Tuke    
2          Edward L. Buck - (I)        1127         Edward L. Buck    
3   Christopher G. Fenner - (I)         230  Christopher G. Fenner    
4     Daniel Towers Lewis - (I)         218    Daniel Towers Lewis    
5              Ed Lawhorn - (I)         207             Ed Lawhorn    
6              Chris Lugo - (I)         195             Chris Lugo    
7          David Gatchell - (I)         145         David Gatchell    
8         Lamar Alexander - (R)       68258        Lamar Alexander    
9          Robert D. Tuke - (D)       19405         Robert D. Tuke    
10         Edward L. Buck - (I)        1262         Edward L. Buck    
11             Chris Lugo - (I)         342             Chris Lugo    
12    Daniel Towers Lewis - (I)         331    Daniel Towers Lewis    
13  Christopher G. Fenner - (I)         289  Christopher G. Fenner    
14             Ed Lawhorn - (I)         277             Ed Lawhorn    
15         David Gatchell - (I)         258         David Gatchell    
16        Lamar Alexander - (R)        5969        Lamar Alexander    
17         Robert D. Tuke - (D)        2620         Robert D. Tuke    
18         Edward L. Buck - (I)         231         Edward L. Buck    
19  Christopher G. Fenner - (I)          54  Christopher G. Fenner    

          party  sum_county_vote  
0    Republican            49143  
1    Democratic            49143  
2           (I)            49143  
3           (I)            49143  
4           (I)            49143  
5           (I)            49143  
6           (I)            49143  
7           (I)            49143  
8    Republican            90422  
9    Democratic            90422  
10          (I)            90422  
11          (I)            90422  
12          (I)            90422  
13          (I)            90422  
14          (I)            90422  
15          (I)            90422  
16   Republican             9016  
17   Democratic             9016  
18          (I)             9016  
19          (I)             9016

In [41]:
# create "pct_county_vote" column for tn_2016_pres
tn_2008_uss['pct_county_vote']=round(tn_2008_uss['vote_count']/tn_2008_uss['sum_county_vote']*100,2)
tn_2008_uss.head(20)

electdate            officename      county  \
0   November 4, 2008  United States Senate      Wilson   
1   November 4, 2008  United States Senate      Wilson   
2   November 4, 2008  United States Senate      Wilson   
3   November 4, 2008  United States Senate      Wilson   
4   November 4, 2008  United States Senate      Wilson   
5   November 4, 2008  United States Senate      Wilson   
6   November 4, 2008  United States Senate      Wilson   
7   November 4, 2008  United States Senate      Wilson   
8   November 4, 2008  United States Senate  Williamson   
9   November 4, 2008  United States Senate  Williamson   
10  November 4, 2008  United States Senate  Williamson   
11  November 4, 2008  United States Senate  Williamson   
12  November 4, 2008  United States Senate  Williamson   
13  November 4, 2008  United States Senate  Williamson   
14  November 4, 2008  United States Senate  Williamson   
15  November 4, 2008  United States Senate  Williamson   
16  November 4, 2008  United States Senate       White   
17  November 4, 2008  United States Senate       White   
18  November 4, 2008  United States Senate       White   
19  November 4, 2008  United States Senate       White   

                      candidate  vote_count                    name  \
0         Lamar Alexander - (R)       34781        Lamar Alexander    
1          Robert D. Tuke - (D)       12240         Robert D. Tuke    
2          Edward L. Buck - (I)        1127         Edward L. Buck    
3   Christopher G. Fenner - (I)         230  Christopher G. Fenner    
4     Daniel Towers Lewis - (I)         218    Daniel Towers Lewis    
5              Ed Lawhorn - (I)         207             Ed Lawhorn    
6              Chris Lugo - (I)         195             Chris Lugo    
7          David Gatchell - (I)         145         David Gatchell    
8         Lamar Alexander - (R)       68258        Lamar Alexander    
9          Robert D. Tuke - (D)       19405         Robert D. Tuke    
10         Edward L. Buck - (I)        1262         Edward L. Buck    
11             Chris Lugo - (I)         342             Chris Lugo    
12    Daniel Towers Lewis - (I)         331    Daniel Towers Lewis    
13  Christopher G. Fenner - (I)         289  Christopher G. Fenner    
14             Ed Lawhorn - (I)         277             Ed Lawhorn    
15         David Gatchell - (I)         258         David Gatchell    
16        Lamar Alexander - (R)        5969        Lamar Alexander    
17         Robert D. Tuke - (D)        2620         Robert D. Tuke    
18         Edward L. Buck - (I)         231         Edward L. Buck    
19  Christopher G. Fenner - (I)          54  Christopher G. Fenner    

          party  sum_county_vote  pct_county_vote  
0    Republican            49143            70.78  
1    Democratic            49143            24.91  
2           (I)            49143             2.29  
3           (I)            49143             0.47  
4           (I)            49143             0.44  
5           (I)            49143             0.42  
6           (I)            49143             0.40  
7           (I)            49143             0.30  
8    Republican            90422            75.49  
9    Democratic            90422            21.46  
10          (I)            90422             1.40  
11          (I)            90422             0.38  
12          (I)            90422             0.37  
13          (I)            90422             0.32  
14          (I)            90422             0.31  
15          (I)            90422             0.29  
16   Republican             9016            66.20  
17   Democratic             9016            29.06  
18          (I)             9016             2.56  
19          (I)             9016             0.60

In [42]:
# how many parties?
tn_2008_uss['party'].value_counts()

 (I)           570
 Republican     95
 Democratic     95
Name: party, dtype: int64

## Create dataframe "tn_2008_pct_county"
_pct"  with percent values per party per county

In [43]:
# create new dataframe with reduced columns values only
tn_2008_pct_county=tn_2008_uss.loc[0:,['electdate','county','party','pct_county_vote']]
tn_2008_pct_county

electdate    county        party  pct_county_vote
0    November 4, 2008    Wilson   Republican            70.78
1    November 4, 2008    Wilson   Democratic            24.91
2    November 4, 2008    Wilson          (I)             2.29
3    November 4, 2008    Wilson          (I)             0.47
4    November 4, 2008    Wilson          (I)             0.44
..                ...       ...          ...              ...
755  November 4, 2008  Anderson          (I)             0.50
756  November 4, 2008  Anderson          (I)             0.39
757  November 4, 2008  Anderson          (I)             0.39
758  November 4, 2008  Anderson          (I)             0.30
759  November 4, 2008  Anderson          (I)             0.28

[760 rows x 4 columns]

In [44]:
# create new dataframe with Republican values only
tn_2008_rep_pct=tn_2008_pct_county[tn_2008_pct_county['party']== " Republican"].reset_index(drop=True)
tn_2008_rep_pct

electdate      county        party  pct_county_vote
0   November 4, 2008      Wilson   Republican            70.78
1   November 4, 2008  Williamson   Republican            75.49
2   November 4, 2008       White   Republican            66.20
3   November 4, 2008     Weakley   Republican            66.69
4   November 4, 2008       Wayne   Republican            76.18
..               ...         ...          ...              ...
90  November 4, 2008      Blount   Republican            79.16
91  November 4, 2008     Bledsoe   Republican            68.88
92  November 4, 2008      Benton   Republican            59.71
93  November 4, 2008     Bedford   Republican            70.37
94  November 4, 2008    Anderson   Republican            71.40

[95 rows x 4 columns]

In [45]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2008_rep_pct=tn_2008_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2008_rep_pct

electdate      county  Republican
0   November 4, 2008      Wilson       70.78
1   November 4, 2008  Williamson       75.49
2   November 4, 2008       White       66.20
3   November 4, 2008     Weakley       66.69
4   November 4, 2008       Wayne       76.18
..               ...         ...         ...
90  November 4, 2008      Blount       79.16
91  November 4, 2008     Bledsoe       68.88
92  November 4, 2008      Benton       59.71
93  November 4, 2008     Bedford       70.37
94  November 4, 2008    Anderson       71.40

[95 rows x 3 columns]

In [46]:
# create new dataframe with Democratic values only
tn_2008_dem_pct=tn_2008_pct_county[tn_2008_pct_county['party']== " Democratic"].reset_index(drop=True)
tn_2008_dem_pct

electdate      county        party  pct_county_vote
0   November 4, 2008      Wilson   Democratic            24.91
1   November 4, 2008  Williamson   Democratic            21.46
2   November 4, 2008       White   Democratic            29.06
3   November 4, 2008     Weakley   Democratic            29.73
4   November 4, 2008       Wayne   Democratic            21.40
..               ...         ...          ...              ...
90  November 4, 2008      Blount   Democratic            17.60
91  November 4, 2008     Bledsoe   Democratic            27.77
92  November 4, 2008      Benton   Democratic            36.63
93  November 4, 2008     Bedford   Democratic            25.70
94  November 4, 2008    Anderson   Democratic            25.65

[95 rows x 4 columns]

In [47]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2008_dem_pct=tn_2008_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2008_dem_pct

electdate      county  Democratic
0   November 4, 2008      Wilson       24.91
1   November 4, 2008  Williamson       21.46
2   November 4, 2008       White       29.06
3   November 4, 2008     Weakley       29.73
4   November 4, 2008       Wayne       21.40
..               ...         ...         ...
90  November 4, 2008      Blount       17.60
91  November 4, 2008     Bledsoe       27.77
92  November 4, 2008      Benton       36.63
93  November 4, 2008     Bedford       25.70
94  November 4, 2008    Anderson       25.65

[95 rows x 3 columns]

In [48]:
# create new dataframe with non Republican or Democratic values only
tn_2008_other_pct=tn_2008_pct_county[(tn_2008_pct_county['party']!=" Republican")&(tn_2008_pct_county['party']!=" Democratic")].reset_index(drop=True)
tn_2008_other_pct

electdate    county party  pct_county_vote
0    November 4, 2008    Wilson   (I)             2.29
1    November 4, 2008    Wilson   (I)             0.47
2    November 4, 2008    Wilson   (I)             0.44
3    November 4, 2008    Wilson   (I)             0.42
4    November 4, 2008    Wilson   (I)             0.40
..                ...       ...   ...              ...
565  November 4, 2008  Anderson   (I)             0.50
566  November 4, 2008  Anderson   (I)             0.39
567  November 4, 2008  Anderson   (I)             0.39
568  November 4, 2008  Anderson   (I)             0.30
569  November 4, 2008  Anderson   (I)             0.28

[570 rows x 4 columns]

In [49]:
# sum by county
tn_2008_other_pct=tn_2008_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2008_other_pct

county  pct_county_vote
0     Anderson             2.94
1      Bedford             3.94
2       Benton             3.67
3      Bledsoe             3.34
4       Blount             3.23
..         ...              ...
90       Wayne             2.42
91     Weakley             3.60
92       White             4.73
93  Williamson             3.07
94      Wilson             4.32

[95 rows x 2 columns]

In [50]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2008_other_pct=tn_2008_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2008_other_pct

county  Other
0     Anderson   2.94
1      Bedford   3.94
2       Benton   3.67
3      Bledsoe   3.34
4       Blount   3.23
..         ...    ...
90       Wayne   2.42
91     Weakley   3.60
92       White   4.73
93  Williamson   3.07
94      Wilson   4.32

[95 rows x 2 columns]

In [53]:
# join tn_2012_rep, tn_2012_dem_pct and tn_2012_other_pct to form tn_2012_county_uss_pct
dem=tn_2008_dem_pct['Democratic']
other=tn_2008_other_pct['Other']
tn_2008_county_uss_pct=tn_2008_rep_pct.join(dem).join(other)
tn_2008_county_uss_pct

electdate      county  Republican  Democratic  Other
0   November 4, 2008      Wilson       70.78       24.91   2.94
1   November 4, 2008  Williamson       75.49       21.46   3.94
2   November 4, 2008       White       66.20       29.06   3.67
3   November 4, 2008     Weakley       66.69       29.73   3.34
4   November 4, 2008       Wayne       76.18       21.40   3.23
..               ...         ...         ...         ...    ...
90  November 4, 2008      Blount       79.16       17.60   2.42
91  November 4, 2008     Bledsoe       68.88       27.77   3.60
92  November 4, 2008      Benton       59.71       36.63   4.73
93  November 4, 2008     Bedford       70.37       25.70   3.07
94  November 4, 2008    Anderson       71.40       25.65   4.32

[95 rows x 5 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [54]:
tn_2008_county_uss_pct['Rep_Margin']= tn_2008_county_uss_pct['Republican']-tn_2008_county_uss_pct['Democratic']
tn_2008_county_uss_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 4, 2008      Wilson       70.78       24.91   2.94       45.87
1   November 4, 2008  Williamson       75.49       21.46   3.94       54.03
2   November 4, 2008       White       66.20       29.06   3.67       37.14
3   November 4, 2008     Weakley       66.69       29.73   3.34       36.96
4   November 4, 2008       Wayne       76.18       21.40   3.23       54.78
5   November 4, 2008  Washington       75.92       21.32   2.71       54.60
6   November 4, 2008      Warren       64.02       31.49   3.74       32.53
7   November 4, 2008   Van Buren       62.27       33.66   6.07       28.61
8   November 4, 2008       Union       75.82       20.20   3.42       55.62
9   November 4, 2008      Unicoi       78.35       18.76   3.22       59.59
10  November 4, 2008   Trousdale       62.75       32.38   3.55       30.37
11  November 4, 2008      Tipton       74.22       22.92   3.43       51.30
12  November 4, 2008      Sumner       71.84       24.19   3.42       47.65
13  November 4, 2008    Sullivan       77.15       20.86   3.92       56.29
14  November 4, 2008     Stewart       59.87       35.43   2.88       24.44
15  November 4, 2008       Smith       64.16       30.74   4.24       33.42
16  November 4, 2008      Shelby       50.88       46.61   2.87        4.27
17  November 4, 2008      Sevier       79.52       16.93   3.79       62.59
18  November 4, 2008  Sequatchie       71.66       24.97   3.57       46.69
19  November 4, 2008       Scott       77.74       18.98   3.02       58.76
20  November 4, 2008  Rutherford       65.44       30.28   4.75       35.16
21  November 4, 2008   Robertson       69.49       26.80   4.41       42.69
22  November 4, 2008       Roane       74.55       22.62   3.62       51.93
23  November 4, 2008        Rhea       76.67       20.38   2.28       56.29
24  November 4, 2008      Putnam       67.38       28.32   3.65       39.06
25  November 4, 2008        Polk       67.83       28.75   3.62       39.08
26  November 4, 2008     Pickett       73.55       23.43   3.33       50.12
27  November 4, 2008       Perry       58.04       38.20   4.47       19.84
28  November 4, 2008     Overton       62.19       34.03   3.62       28.16
29  November 4, 2008       Obion       68.18       28.65   2.57       39.53
30  November 4, 2008      Morgan       75.86       21.32   3.66       54.54
31  November 4, 2008       Moore       70.01       26.43   2.65       43.58
32  November 4, 2008  Montgomery       62.65       33.01   2.10       29.64
33  November 4, 2008      Monroe       75.18       21.83   3.31       53.35
34  November 4, 2008       Meigs       72.33       24.50   3.08       47.83
35  November 4, 2008     Mcnairy       69.29       27.91   3.08       41.38
36  November 4, 2008      Mcminn       75.35       21.45   3.11       53.90
37  November 4, 2008       Maury       61.94       34.59   1.81       27.35
38  November 4, 2008    Marshall       64.59       30.69   3.28       33.90
39  November 4, 2008      Marion       62.18       34.76   2.93       27.42
40  November 4, 2008     Madison       59.15       37.96   5.01       21.19
41  November 4, 2008       Macon       72.22       21.40   3.50       50.82
42  November 4, 2008      Loudon       80.28       17.16   3.50       63.12
43  November 4, 2008     Lincoln       69.94       25.84   4.99       44.10
44  November 4, 2008       Lewis       63.73       31.92   3.66       31.81
45  November 4, 2008    Lawrence       70.50       26.21   4.24       44.29
46  November 4, 2008  Lauderdale       64.58       32.31   3.03       32.27
47  November 4, 2008        Lake       56.47       37.27   6.25       19.20
48  November 4, 2008        Knox       71.22       25.75   3.11       45.47
49  November 4, 2008     Johnson       76.12       19.62   3.27       56.50

In [55]:
# correct county names: Dekalb to DeKalb, Mcminn to McMinn, Mcnairy to McNairy 
tn_2008_county_uss_pct['county']=tn_2008_county_uss_pct['county'].str.replace('Dekalb','DeKalb').str.replace('Mcminn','McMinn').str.replace('Mcnairy','McNairy')
tn_2008_county_uss_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 4, 2008      Wilson       70.78       24.91   2.94       45.87
1   November 4, 2008  Williamson       75.49       21.46   3.94       54.03
2   November 4, 2008       White       66.20       29.06   3.67       37.14
3   November 4, 2008     Weakley       66.69       29.73   3.34       36.96
4   November 4, 2008       Wayne       76.18       21.40   3.23       54.78
5   November 4, 2008  Washington       75.92       21.32   2.71       54.60
6   November 4, 2008      Warren       64.02       31.49   3.74       32.53
7   November 4, 2008   Van Buren       62.27       33.66   6.07       28.61
8   November 4, 2008       Union       75.82       20.20   3.42       55.62
9   November 4, 2008      Unicoi       78.35       18.76   3.22       59.59
10  November 4, 2008   Trousdale       62.75       32.38   3.55       30.37
11  November 4, 2008      Tipton       74.22       22.92   3.43       51.30
12  November 4, 2008      Sumner       71.84       24.19   3.42       47.65
13  November 4, 2008    Sullivan       77.15       20.86   3.92       56.29
14  November 4, 2008     Stewart       59.87       35.43   2.88       24.44
15  November 4, 2008       Smith       64.16       30.74   4.24       33.42
16  November 4, 2008      Shelby       50.88       46.61   2.87        4.27
17  November 4, 2008      Sevier       79.52       16.93   3.79       62.59
18  November 4, 2008  Sequatchie       71.66       24.97   3.57       46.69
19  November 4, 2008       Scott       77.74       18.98   3.02       58.76
20  November 4, 2008  Rutherford       65.44       30.28   4.75       35.16
21  November 4, 2008   Robertson       69.49       26.80   4.41       42.69
22  November 4, 2008       Roane       74.55       22.62   3.62       51.93
23  November 4, 2008        Rhea       76.67       20.38   2.28       56.29
24  November 4, 2008      Putnam       67.38       28.32   3.65       39.06
25  November 4, 2008        Polk       67.83       28.75   3.62       39.08
26  November 4, 2008     Pickett       73.55       23.43   3.33       50.12
27  November 4, 2008       Perry       58.04       38.20   4.47       19.84
28  November 4, 2008     Overton       62.19       34.03   3.62       28.16
29  November 4, 2008       Obion       68.18       28.65   2.57       39.53
30  November 4, 2008      Morgan       75.86       21.32   3.66       54.54
31  November 4, 2008       Moore       70.01       26.43   2.65       43.58
32  November 4, 2008  Montgomery       62.65       33.01   2.10       29.64
33  November 4, 2008      Monroe       75.18       21.83   3.31       53.35
34  November 4, 2008       Meigs       72.33       24.50   3.08       47.83
35  November 4, 2008     McNairy       69.29       27.91   3.08       41.38
36  November 4, 2008      McMinn       75.35       21.45   3.11       53.90
37  November 4, 2008       Maury       61.94       34.59   1.81       27.35
38  November 4, 2008    Marshall       64.59       30.69   3.28       33.90
39  November 4, 2008      Marion       62.18       34.76   2.93       27.42
40  November 4, 2008     Madison       59.15       37.96   5.01       21.19
41  November 4, 2008       Macon       72.22       21.40   3.50       50.82
42  November 4, 2008      Loudon       80.28       17.16   3.50       63.12
43  November 4, 2008     Lincoln       69.94       25.84   4.99       44.10
44  November 4, 2008       Lewis       63.73       31.92   3.66       31.81
45  November 4, 2008    Lawrence       70.50       26.21   4.24       44.29
46  November 4, 2008  Lauderdale       64.58       32.31   3.03       32.27
47  November 4, 2008        Lake       56.47       37.27   6.25       19.20
48  November 4, 2008        Knox       71.22       25.75   3.11       45.47
49  November 4, 2008     Johnson       76.12       19.62   3.27       56.50

In [56]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [57]:
# join with tn_2012_county_uss_pct
tn_2008_county_uss_pct=tn_2008_county_uss_pct.merge(county_index,on="county")
tn_2008_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  \
0   November 4, 2008      Wilson       70.78       24.91   2.94       45.87   
1   November 4, 2008  Williamson       75.49       21.46   3.94       54.03   
2   November 4, 2008       White       66.20       29.06   3.67       37.14   
3   November 4, 2008     Weakley       66.69       29.73   3.34       36.96   
4   November 4, 2008       Wayne       76.18       21.40   3.23       54.78   
..               ...         ...         ...         ...    ...         ...   
90  November 4, 2008      Blount       79.16       17.60   2.42       61.56   
91  November 4, 2008     Bledsoe       68.88       27.77   3.60       41.11   
92  November 4, 2008      Benton       59.71       36.63   4.73       23.08   
93  November 4, 2008     Bedford       70.37       25.70   3.07       44.67   
94  November 4, 2008    Anderson       71.40       25.65   4.32       45.75   

    INDEX_NO  
0         65  
1         44  
2         67  
3         56  
4          6  
..       ...  
90        52  
91        31  
92        41  
93        26  
94        79  

[95 rows x 7 columns]

In [58]:
tn_2008_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   county      95 non-null     object 
 2   Republican  95 non-null     float64
 3   Democratic  95 non-null     float64
 4   Other       95 non-null     float64
 5   Rep_Margin  95 non-null     float64
 6   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 5.9+ KB


In [59]:
# covert "electdate" to datetime from object
tn_2008_county_uss_pct['electdate']=pd.to_datetime(tn_2008_county_uss_pct['electdate'],format='%B %d, %Y')
tn_2008_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  INDEX_NO
0  2008-11-04      Wilson       70.78       24.91   2.94       45.87        65
1  2008-11-04  Williamson       75.49       21.46   3.94       54.03        44
2  2008-11-04       White       66.20       29.06   3.67       37.14        67
3  2008-11-04     Weakley       66.69       29.73   3.34       36.96        56
4  2008-11-04       Wayne       76.18       21.40   3.23       54.78         6
..        ...         ...         ...         ...    ...         ...       ...
90 2008-11-04      Blount       79.16       17.60   2.42       61.56        52
91 2008-11-04     Bledsoe       68.88       27.77   3.60       41.11        31
92 2008-11-04      Benton       59.71       36.63   4.73       23.08        41
93 2008-11-04     Bedford       70.37       25.70   3.07       44.67        26
94 2008-11-04    Anderson       71.40       25.65   4.32       45.75        79

[95 rows x 7 columns]

In [79]:
tn_2008_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   county      95 non-null     object        
 2   Republican  95 non-null     float64       
 3   Democratic  95 non-null     float64       
 4   Other       95 non-null     float64       
 5   Rep_Margin  95 non-null     float64       
 6   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [60]:
tn_2008_county_uss_pct.to_excel(r'../data/2008_election/tn_2008_county_uss_pct.xlsx',index=False)